### PR TITLE
Ek cramreference 677

### PR DESCRIPTION
**Pull request by edwardkw**

### CODE REVIEW

File: src/java/htsjdk/samtools/Defaults.java [hunk 1]
```diff
@@ -73,6 +79,7 @@
             NON_ZERO_BUFFER_SIZE = BUFFER_SIZE;
         }
         REFERENCE_FASTA = getFileProperty("reference_fasta", null);
+        REF_DOWNLOAD_ENABLED = false;
```
Comment (droazen) [inline src/java/htsjdk/samtools/Defaults.java]: You need to call `getBooleanProperty("ref_download_enabled", false)` here to allow the default of false to be overridden by a system property.

Comment (droazen) [inline src/java/htsjdk/samtools/Defaults.java]: Also, the property name should probably make it clear that this controls a CRAM-specific feature.

File: src/java/htsjdk/samtools/cram/ref/ReferenceSource.java [hunk 2]
```diff
@@ -108,7 +108,7 @@ public void clearCache() {
             }
         }
 
-        { // try to fetch sequence by md5:
+        if (Defaults.REF_DOWNLOAD_ENABLED) { // try to fetch sequence by md5:
             if (md5 != null)
```
Comment (nh13) [inline src/java/htsjdk/samtools/cram/ref/ReferenceSource.java]: Missing ```{``` and ```}```.  Could you fix here since now we have nested ```if```s?

**Comment by edwardkw** [comment]
Closed because this isn't meant to be merged yet.

File: src/java/htsjdk/samtools/Defaults.java [hunk 3]
```diff
@@ -53,6 +53,12 @@
     public static final String CUSTOM_READER_FACTORY;
 
     /**
+     * Boolean describing whether downloads of reference files are allowed,
+     * given that the reference file is not specified by the user
+     */
+    public static final boolean CRAM_REF_DOWNLOAD_ENABLED;
```
Comment (nh13) [inline src/java/htsjdk/samtools/Defaults.java]: Minor: How about ```USE_CRAM_REF_DOWNLOAD``` to be consistent with ```USE``` elsewhere?

File: src/java/htsjdk/samtools/Defaults.java [hunk 4]
```diff
@@ -53,6 +53,12 @@
     public static final String CUSTOM_READER_FACTORY;
 
     /**
+     * Boolean describing whether downloads of reference files are allowed,
```
Comment (nh13) [inline src/java/htsjdk/samtools/Defaults.java]: should state for CRAM files only.

File: src/java/htsjdk/samtools/cram/ref/ReferenceSource.java [hunk 5]
```diff
@@ -108,17 +108,19 @@ public void clearCache() {
...
 
-        { // try to fetch sequence by md5:
-            if (md5 != null)
-                try {
-                    bases = findBasesByMD5(md5.toLowerCase());
-                } catch (final Exception e) {
-                    throw new RuntimeException(e);
+        {
+            if (Defaults.CRAM_REF_DOWNLOAD_ENABLED) { // try to fetch sequence by md5:
+                if (md5 != null)
```
Comment (nh13) [inline src/java/htsjdk/samtools/cram/ref/ReferenceSource.java]: Missing ```{...}```.

**Comment by nh13** [comment]
Minor comments, :+1:

**Comment by vadimzalunin** [comment]
+1

**Comment by edwardkw** [comment]
If there's no other issues, merge when ready. Thanks!

File: src/java/htsjdk/samtools/Defaults.java [hunk 6]
```diff
@@ -73,6 +80,7 @@
             NON_ZERO_BUFFER_SIZE = BUFFER_SIZE;
         }
         REFERENCE_FASTA = getFileProperty("reference_fasta", null);
+        USE_CRAM_REF_DOWNLOAD = getBooleanProperty("ref_download_enabled", false);
```
Comment (droazen) [inline src/java/htsjdk/samtools/Defaults.java]: Can the property name that you pass to `getBooleanProperty()` match the variable name? Eg.,

    USE_CRAM_REF_DOWNLOAD = getBooleanProperty("use_cram_ref_download", false);

**Comment by droazen** [comment]
One final request for @edwardkw before merge.

**Comment by droazen** [comment]
:+1:

**Pull request merged**